transform-plugins/src/main/java/co/cask/hydrator/plugin/JSONParser.java

```diff
@@ -61,21 +80,85 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) throws Ille
+  // populate the output schema fields.
+  //
+  // E.g. expensive:$.expensive maps the input Json path from root, field expensive to expensive.
+  private void extractMappings() {
+    if (config.mapping != null && config.mapping.isEmpty()) {
+      isSimple = true;
+    } else {
+      isSimple = false;
+      String[] pathMaps = config.mapping.split(",");
+      for (String pathMap : pathMaps) {
+        String[] mapParts = pathMap.split(":");
+        String field = mapParts[0];
+        String expression = mapParts[1];
+        if (field.isEmpty() && !expression.isEmpty()) {
+          throw new IllegalArgumentException("Json path expression '" + expression +
+                  "' has not output field specified");
+        }
+        if (expression.isEmpty() && !field.isEmpty()) {
+          throw new IllegalArgumentException("Field '" + field + "' doesn't have Json path expression.");
+        }
+        mapping.put(field, expression);
+      }
+    }
   }
-  
+
   @Override
   public void initialize(TransformContext context) throws Exception {
     super.initialize(context);
     try {
       outSchema = Schema.parseJson(config.schema);
+      fields = outSchema.getFields();
     } catch (IOException e) {
       throw new IllegalArgumentException("Output Schema specified is not a valid JSON. Please check the Schema JSON");
     }
+    extractMappings();
   }
-  
+
   @Override
   public void transform(StructuredRecord input, Emitter<StructuredRecord> emitter) throws Exception {
-    emitter.emit(StructuredRecordStringConverter.fromJsonString((String) input.get(config.field), outSchema));
+    // If it's a simple mapping from JSON to output schema, else we use the mapping fields to map the
+    // the JSON using JSON path to fields. This is used for mapping complex JSON schemas.
+    if (isSimple) {
+      emitter.emit(StructuredRecordStringConverter.fromJsonString((String) input.get(config.field), outSchema));
+      return;
+    }
+
+    // When it's not a simple Json to be parsed, we use the Json path to map the input Json fields into the
+    // output schema. In order to optimize for reading multiple paths from the Json we create a document that
+    // allows the Json to be parsed only once. We then iterate through the output fields and apply the
+    // path to extract the fields.
+    Object document = Configuration.defaultConfiguration().jsonProvider().parse((String) input.get(config.field));
+    StructuredRecord.Builder builder = StructuredRecord.builder(outSchema);
+    for (Schema.Field field : fields) {
+      String name = field.getName();
+      if (mapping.containsKey(name)) {
+        String path = mapping.get(name);
+        try {
+          Object value = JsonPath.read(document, path);
+          builder.set(field.getName(), value);
+        } catch (PathNotFoundException e) {
+          LOG.error("Json path '" + path + "' specified for the field '" + name + "' doesn't exist. " +
```

Comment:
Should we also include there variable `e` as the last parameter of the LOG.error method? So that the exception is also included in the error log.